test/PspUpf.jl

```diff
@@ -0,0 +1,125 @@
+using DFTK: parse_upf_file
+using SpecialFunctions: sphericalbesselj
+using QuadGK
+
+hgh_upf_files = ["si-q4", "tl-q13"]
+oncv_upf_files = ["si-nc-sr-standard-04", "hf-sp-oncvpsp"]
+all_upf_files = vcat(hgh_upf_files, oncv_upf_files)
+
+@testset "Check reading 'si-nc-sr-standard-04.upf'" begin
+    psp = parse_upf_file("psp/si-nc-sr-standard-04.upf")
+
+    @test psp.lmax == 2
+    @test psp.Zion == 4
+    @test length(psp.rgrid) == 1510
+    @test length(psp.vloc) == 1510
+    for m in psp.h
+        @test size(m) == (2, 2)
+    end
+
+    @test psp.projs[1][1][1] ≈ -5.6328824383E-09 / 2
+end
+
+@testset "Real potentials are consistent with HGH" begin
+    for pspfile in hgh_upf_files
+        upf = parse_upf_file("psp/$(pspfile).upf")
+        hgh = load_psp("hgh/pbe/$(pspfile).hgh")
+        for r in [upf.rgrid[1], upf.rgrid[end]]
+            reference_hgh = eval_psp_local_real(hgh, r)
+            @test reference_hgh ≈ eval_psp_local_real(upf, r) rtol=1e-8 atol=1e-8
+        end
+    end
+end
+
+@testset "Fourier potentials are consistent with HGH" begin
+    for pspfile in hgh_upf_files
+        upf = parse_upf_file("psp/$(pspfile).upf")
+        hgh = load_psp("hgh/pbe/$(pspfile).hgh")
+        for q in [0.01, 0.1, 0.2, 0.5, 1., 2., 5., 10.]
+            reference_hgh = eval_psp_local_fourier(hgh, q)
+            @test reference_hgh ≈ eval_psp_local_fourier(upf, q) rtol=1e-5 atol=1e-5
+        end
+    end
+end
+
+@testset "Projectors are consistent with HGH in real and Fourier space" begin
+    for pspfile in hgh_upf_files
+        upf = parse_upf_file("psp/$(pspfile).upf")
+        hgh = load_psp("hgh/pbe/$(pspfile).hgh")
+        for (l, i) in [(0, 1), (0, 2), (0, 3), (1, 1), (1, 2), (1, 3),
+                       (2, 1), (2, 2), (3, 1)]
+            l > upf.lmax - 1 && continue  # Overshooting available AM
+            i > length(upf.projs[l+1]) && continue  # Overshooting available projectors at AM
+            ircut = length(upf.projs[l+1][i])
+            for q in [0.01, 0.1, 0.2, 0.5, 1., 2., 5., 10.]
+                reference_hgh = eval_psp_projector_fourier(hgh, i, l, q)
+                @test reference_hgh ≈ eval_psp_projector_fourier(upf, i, l, q) atol=1e-7 rtol=1e-7
+            end
+            for r in [upf.rgrid[1], upf.rgrid[ircut]]
+                reference_hgh = eval_psp_projector_real(hgh, i, l, r)
+                @test reference_hgh ≈ eval_psp_projector_real(upf, i, l, r) atol=1e-7 rtol=1e-7
+            end
+        end
+    end
+end
+
+@testset "Energy correction is consistent with HGH" begin
+    for pspfile in hgh_upf_files
+        upf = parse_upf_file("psp/$(pspfile).upf")
+        hgh = load_psp("hgh/pbe/$(pspfile).hgh")
+        n_electrons = 3
+        reference_hgh = eval_psp_energy_correction(hgh, n_electrons)
+        @test reference_hgh ≈ eval_psp_energy_correction(upf, n_electrons) rtol=1e-5 atol=1e-5
+    end
+end
+
+@testset "Potentials are consistent in real and Fourier space" begin
```

Comment:
The integrand functions + quadgk could be used to define a default for evaluating the FT of the potential and projectors for `NormConservingPsp`s. Otherwise, what you suggest would be good.

If we're planning to move out the pseudo stuff to `PseudoPotentalsIO.jl`, then we could just wait until that point to do some reorganization.

Reply:
I was mostly referring to the `n_proj` stuff (so all good what you did), but I agree that the numerical integration routines should also have a generic fallback.